libraries/abstractions/secure_sockets/test/iot_test_udp.c

```diff
@@ -479,32 +478,107 @@ static BaseType_t prvCheckRxTxBuffers( uint8_t * pucTxBuffer,
         {
             xReturn = pdFAIL;
             udptestFAILUREPRINTF( ( "Message bytes received different than those sent. Message Length %d, Byte Index %d \r\n", xMessageLength, xIndex ) );
-            udptestASSERT( xReturn == pdPASS );
+            break;
         }
     }
 
-    /* Check that neither the Rx nor Tx buffers were modified/overflowed. */
-    for( xIndex = xMessageLength; xIndex < udptestBUFFER_SIZE; xIndex++ )
+    if( xReturn == pdPASS )
     {
-        if( pucRxBuffer[ xIndex ] != udptestRX_BUFFER_FILLER )
+        /* Check that neither the Rx nor Tx buffers were modified/overflowed. */
+        for( xIndex = xMessageLength; xIndex < udptestBUFFER_SIZE; xIndex++ )
         {
-            xReturn = pdFAIL;
-            udptestFAILUREPRINTF( ( "Rx buffer padding modified byte. Message Length %d, Byte Index %d \r\n", xMessageLength, xIndex ) );
-        }
+            if( pucRxBuffer[ xIndex ] != udptestRX_BUFFER_FILLER )
+            {
+                xReturn = pdFAIL;
+                udptestFAILUREPRINTF( ( "Rx buffer padding modified byte. Message Length %d, Byte Index %d \r\n", xMessageLength, xIndex ) );
+            }
 
-        if( pucTxBuffer[ xIndex ] != udptestTX_BUFFER_FILLER )
-        {
-            xReturn = pdFAIL;
-            udptestFAILUREPRINTF( ( "Tx buffer padding modified. Message Length %d, Byte Index %d \r\n", xMessageLength, xIndex ) );
+            if( pucTxBuffer[ xIndex ] != udptestTX_BUFFER_FILLER )
+            {
+                xReturn = pdFAIL;
+                udptestFAILUREPRINTF( ( "Tx buffer padding modified. Message Length %d, Byte Index %d \r\n", xMessageLength, xIndex ) );
+            }
         }
     }
 
-    udptestASSERT( xReturn == pdPASS );
+    /* udptestASSERT( xReturn == pdPASS ); */
 
     return xReturn;
 }
 /*-----------------------------------------------------------*/
 
+/*
+ * @brief Test SendTo and RecvFrom echo server. Return pdFAIL if one of the operation fails, allowing retry.
+ * We expect the socket to not fail during retry.
+ */
+static BaseType_t prvSendRecvEchoTestHelper( Socket_t xSocket,
+                                             size_t xLength,
+                                             SocketsSockaddr_t * pxEchoServerAddress,
+                                             uint32_t ulTxCount )
+{
+    BaseType_t xResult;
+    BaseType_t xIntermResult;
+    uint8_t * pucTxBuffer = ( uint8_t * ) pcTxBuffer;
+    uint8_t * pucRxBuffer = ( uint8_t * ) pcRxBuffer;
+    SocketsSockaddr_t xReceiveAddress = { 0 };
+    uint32_t ulRetry;
+
+    xIntermResult = pdPASS;
+    xResult = pdFAIL;
+    ulRetry = 0;
+
+    memset( pucTxBuffer, udptestTX_BUFFER_FILLER, udptestBUFFER_SIZE );
+
+    prvCreateTxData( ( char * ) pucTxBuffer,
+                     xLength,
+                     ulTxCount );
+
+    for( ; ulRetry < udptestMAX_RETRY; ulRetry++ )
```

Comment:
Why do we need retries in the test?  Is this test reliable?  How do we know we have enough/too many retries?

Reply:
The reason behind retries should be in the PR description. Adding retries to 2 on TI board in lab setup should be enough that test will only fail due to network reasons once every 4 years of nightly test (using the 2% fail rate per loop). But the fail rate is different on each board, and adding wifi to the equation may increase the fail rate. 

Currently the retry is set to 8. This amount of retries should given us assurance that the if UDP fail it may be the setup or implementation and will not have to worry about network/wifi issue.

Reply:
I have a number of questions on this PR:

1. What is the purpose of this test?  To determine if UDP has been implemented correctly or to determine that it can work some percentage of the time?
1. How do we know that the test is failing due to connectivity or infrastructure problems rather than an intermittent bug in the code?
1. This library should be platform-independent, why are we basing the number of retries in a test on historical results from a particular board (TI) in a particular HW lab?  This test will be run in many different environments and on many different platforms.  Would this number of retries be appropriate for a platform which uses wired Ethernet or PPP over serial for example?

Reply:
1. We are trying to make sure implementation of the socket is correct and works on hardware. If the test is passed, we are confident the device is able to send and receive UDP data, vary in data length.
2. With the same socket code on a device the fail rate varies randomly (from 0.1% to 2%+) when testing at different time. We can try to find the root cause of the random factor, otherwise we are suspecting instability of connection to the server.
3. The library is platform-independent, this test will run on all platforms the same configuration. Tests on software socket libraries (FreeRTOS+TCP and LWIP) has the lowest fail rate passing all 80 tries, but occasionally one or two loops will fail causing the whole test to fail. Hardware implementation of UDP are observed to fail more often. We can adjust the number of retries base on different platform, but may require more formal testing to determine those numbers.